### PR TITLE
[SOLR-17368] TestPrometheusResponseWriter redesign

### PR DESCRIPTION
**Pull request by mlbiscoc**
https://issues.apache.org/jira/browse/SOLR-17368

# Description

Please provide a short description of the changes you're making with this pull request.

The `TestPrometheusResponseWriter` was failing 100% of Jenkins builds due to leaking metrics from other tests running on the same JVM. This PR will redesigns the `TestPrometheusResponseWriter` to only test the structure of the output from the PrometheusResponseWriter and not the output of every metric that exists in Solr.

# Solution

Please provide a short description of the approach taken to implement your solution.

Removed the `solr-prometheus-output.txt` output entirely. Filter out comments from the output (i.e `#` lines) and make an assertion that all metric names must start with `solr_metrics_*` and values must be a valid Prometheus accepted value `floating point number, NaN, +Inf, or -Inf` per Prometheus exposition format documentation. 

Also created `testPrometheusDummyOutput` that registers a dummy metrics for each registry and asserts against it's expected value. Because these are non-existent dummy values, any leaking metrics from other metrics should not override these values.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

`testPrometheusStructureOutput` and `testPrometheusDummyOutput`

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/core/src/test/org/apache/solr/response/TestPrometheusResponseWriter.java [hunk 1]
```diff
@@ -34,57 +38,104 @@
 import org.junit.ClassRule;
 import org.junit.Test;
 
-/** Tests the {@link PrometheusResponseWriter} behavior */
-@LuceneTestCase.AwaitsFix(bugUrl = "https://issues.apache.org/jira/browse/SOLR-17368")
 public class TestPrometheusResponseWriter extends SolrTestCaseJ4 {
-
   @ClassRule public static SolrJettyTestRule solrClientTestRule = new SolrJettyTestRule();
   public static JettySolrRunner jetty;
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/response/TestPrometheusResponseWriter.java:43]: I think this field serves no purpose; we can get a jetty if needed from the rule on the fly.  Static fields in tests can have a maintenance burden to ensure they get null'ed.

File: solr/core/src/test/org/apache/solr/response/TestPrometheusResponseWriter.java [hunk 2]
```diff
@@ -34,57 +38,104 @@
...
+  public static final List<String> VALID_PROMETHEUS_VALUES = Arrays.asList("NaN", "+Inf", "-Inf");
 
   @BeforeClass
   public static void beforeClass() throws Exception {
     solrClientTestRule.startSolr(LuceneTestCase.createTempDir());
     jetty = solrClientTestRule.getJetty();
     solrClientTestRule.newCollection().withConfigSet(ExternalPaths.DEFAULT_CONFIGSET).create();
     jetty.getCoreContainer().waitForLoadingCoresToFinish(30000);
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/response/TestPrometheusResponseWriter.java:51]: can call solrClientTestRule.getCoreContainer() instead and same below; maybe put in a var "cc" (a popular abbreviation in Solr for this super-common class)

**Review by dsmiley — COMMENTED** [review]
In some sense this is an improvement in less flakiness (yay) but it doesn't address the intent of why solr-prometheus-output.txt was being written in the first place.  Alas; it's a contentious topic and Hossman is opinionated so I accept the PR.

**Comment by dsmiley** [comment]
>  any leaking metrics from other metrics should not override these values.

That's dodging an issue that should be investigated as it might interfere with another metrics based test I suppose.  Any way; progress not perfection.

**Comment by mlbiscoc** [comment]
> > any leaking metrics from other metrics should not override these values.
> 
> That's dodging an issue that should be investigated as it might interfere with another metrics based test I suppose. Any way; progress not perfection.

If I get some time, I can take a look into the lifecycle issue then we can revisit `solr-prometheus-output.txt`

**Pull request merged**